File: adapters/ui-swift/Sources/FluxUIKit/ButtonAdapter.swift
//  ButtonAdapter.swift
//  FluxUIKit — `Button` → `UIButton` (Appendix F.2).

import UIKit

/// Declarative adapter mapping a Flux `Button` node to a `UIButton`
/// (unified tier; AGENTS.md §3.5).
///
/// Props are read by name; the index is the FNV-1a-32 digest of the name
/// masked to `u16` (`Props.propIndex`), derived identically on server and
/// client (AGENTS.md §3.2) — never a hardcoded positional index. Fields:
/// - `text: String` (required)
/// - `onClick: Handler` (required)
/// - `enabled: Bool = true`
/// - `color: Option[Color]`
///
/// Tapping the button dispatches `onClick` via the weak executor. The target
/// is held only by the button's action, which cannot resurrect a deallocated
/// runtime because it keeps the executor `weak`.
public final class ButtonAdapter: FluxAdapter {
    public typealias View = UIButton
    weak var executor: (any FluxExecutor)?

    public init(executor: (any FluxExecutor)? = nil) { self.executor = executor }

    public func create() -> UIButton {
        UIButton(type: .system)
    }

    public func update(_ view: UIButton, from old: Props, to new: Props) {
        let title = new.getString(named: "text") ?? old.getString(named: "text") ?? ""
        view.setTitle(title, for: .normal)
        view.isEnabled = new.getBool(named: "enabled") ?? true
        if let color = new.getColor(named: "color") { view.setTitleColor(color.uiColor, for: .normal) }
    }

    public func setChildren(_ children: [AnyObject], on view: UIButton) {}

    public func bindHandler(_ handlerId: FluxHandlerId, to view: UIButton, nodeId: FluxNodeId) {
        let target = HandlerTarget(executor: executor, handlerId: handlerId, nodeId: nodeId) { nil }
        view.addAction(UIAction { _ in target.fire() }, for: .touchUpInside)
    }

    public func destroy(_ view: UIButton) {
        view.removeTarget(nil, action: nil, for: .allEvents)
    }
}

File: adapters/ui-kotlin/src/main/kotlin/dev/flux/ui/ButtonAdapter.kt
package dev.flux.ui

import java.lang.ref.WeakReference
import kotlin.collections.List as KList

/**
 * Declarative adapter for `Button` (unified tier; AGENTS.md §3.5).
 *
 * Maps a Flux `Button` node to a native button view. Tapping the view
 * dispatches the `onClick` handler through the weakly-held executor. The
 * handler id is read fresh in [bindHandler] so a hot-swapped closure table is
 * used for the next tap.
 *
 * Each node gets its own adapter instance via [create], so the bound
 * `WeakReference<FluxExecutor>` and handler id never leak into a sibling node
 * (FLUX-007).
 */
public class ButtonAdapter private constructor() : FluxAdapter<FluxNativeView> {
    override val kind: String = KIND

    override fun create(nodeId: UInt): FluxNativeView = FluxNativeViewImpl(nodeId, kind)

    override fun update(
        view: FluxNativeView,
        props: Props,
    ) {
        val text = props.getString(PropsIndex.BUTTON_TEXT).orEmpty()
        if (view.getProperty(PROP_TEXT) != text) view.setProperty(PROP_TEXT, text)

        val enabled = props.getBool(PropsIndex.BUTTON_ENABLED, true)
        if (view.getProperty(PROP_ENABLED) != enabled) view.setProperty(PROP_ENABLED, enabled)

        props.getColor(PropsIndex.BUTTON_COLOR)?.let { color ->
            if (view.getProperty(PROP_COLOR) != color) view.setProperty(PROP_COLOR, color)
        }

        // FLUX-044: surface host-render-only a11y props to the native view.
        view.applyAccessibility(props)
    }

    override fun setChildren(
        view: FluxNativeView,
        childIds: KList<UInt>,
        children: KList<FluxNativeView>,
    ) {
        // Button has no children.
    }

    override fun bindHandler(
        view: FluxNativeView,
        props: Props,
        executor: WeakReference<FluxExecutor>,
    ) {
        val handlerId = props.getHandler(PropsIndex.BUTTON_ON_PRESS)
        view.setProperty(PROP_HANDLER, handlerId)
        view.setProperty(PROP_EXECUTOR, executor)
    }

    override fun destroy(view: FluxNativeView) {
        view.setProperty(PROP_EXECUTOR, null)
        view.setProperty(PROP_HANDLER, 0u)
        view.clearChildren()
    }

    internal companion object {
        /** The kind tag this adapter handles. Exposed for the factory map. */
        const val KIND: String = "button"

        /** Builds a fresh [ButtonAdapter] for one IR node (FLUX-007). */
        fun create(): FluxAdapter<FluxNativeView> = ButtonAdapter()

        const val PROP_TEXT = "text"
        const val PROP_ENABLED = "enabled"
        const val PROP_COLOR = "color"
        const val PROP_HANDLER = "onPressHandler"
        const val PROP_EXECUTOR = "executor"
    }
}
